### PR TITLE
 Changed bundler to Rollup

### LINKED ISSUE #262: ES6 Modules
Is there a real reason, why we don't ship ES6 modules?
I would like to see an additional export at this point, I would also do a PR, if wished and needed.

Regards, 
   Johann

### PR DESCRIPTION
**Pull request by krisk**
Ok, a bunch of goodies here:

- Added ES6 modules for bundlers and browsers (fixed #262)
- Added CommonJS
- Added UMD
- Refactored TS definition, and filled out all its missing parts
- Converted source code to ES6
- Name change so that _.min_ actually reflects the minified version. I bet this will surprise a few folks. Forgive me, world 🙏.

cc: @KonradHoeffner @sidvishnoi @cshoredaniel

### CODE REVIEW

File: src/core/index.js [hunk 1]
```diff
@@ -0,0 +1,284 @@
+
+import { BitapSearch, ExtendedSearch, NGramSearch } from '../search'
+import { isArray, isDefined, isString, isNumber, isObject } from '../helpers/type-checkers'
+import get from '../helpers/get'
+import { createIndex as createIdx, KeyStore } from '../tools'
```
Comment (krisk) [inline src/core/index.js]: Remove rename

File: build/configs.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+const path = require('path')
+const buble = require('@rollup/plugin-buble')
+const replace = require('@rollup/plugin-replace')
+const node = require('@rollup/plugin-node-resolve')
+const babel = require('rollup-plugin-babel')
+const featureFlags = require('./feature-flags')
+const pckg = require('../package.json')
+
+const LIBRARY_NAME = 'Fuse.js'
+const FILENAME = 'fuse'
```
Comment (krisk) [inline build/configs.js]: Unused

File: build/configs.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */`
+
+const resolve = _path => path.resolve(__dirname, '../', _path)
+
+const builds = {
+  // UMD build
+  'umd-dev': {
+    entry: resolve('src/index.js'),
```
Comment (krisk) [inline build/configs.js]: Repeated line.

File: build/build.main.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      if (num < total) {
+        next(num)
+      }
+    } catch (err) {
+      logError(err)
+    }
+  }
+
+  next(0)
+}
```
Comment (sidvishnoi) [inline build/build.main.js]:
```suggestion
async function build(builds) {
  for (const entry of builds) {
    try {
      buildEntry(entry)
    } catch (err) {
      logError(err)
    }
  }
}
```

File: build/build.main.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  const output = config.output
+  const { file, banner } = output
+  const isProd = /(min|prod)\.js$/.test(file)
+
+  try {
+    let bundle = await rollup.rollup(config)
+    let { output: [{ code }] } = await bundle.generate(output)
+
+    if (isProd) {
+      const minified = (banner ? banner + '\n' : '') + terser.minify(code, {
```
Comment (sidvishnoi) [inline build/build.main.js]: Nit: Maybe add the `\n` to banner itself?

File: build/build.main.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    })
+  })
+}
+
+function getSize(code) {
+  return (code.length / 1024).toFixed(2) + 'kb'
+}
+
+function logError(e) {
+  console.log(e)
```
Comment (sidvishnoi) [inline build/build.main.js]:
```suggestion
  console.error(e)
```

**Comment by sidvishnoi** [comment]
I love it! Great work! 
Massive PR by the way! :smile_cat:

File: README.md [hunk 7]
```diff
@@ -43,6 +47,19 @@ $ npm install --save fuse.js
...
+
+Available on CDN via [jsDelivr](https://cdn.jsdelivr.net/npm/fuse.js/dist/).
+
+### Explanation of Different Builds
+
+In the [`dist/` directory of the NPM package](https://cdn.jsdelivr.net/npm/fuse.js/dist/) you will find many different builds of Fuse.js. Here's an overview of the difference between them:
+
+|                       | UMD         | CommonJS       | ES Module (for bundlers) | ES Module (for browsers) |
+| --------------------- | ----------- | -------------- | ------------------------ | ------------------------ |
+| **Full**              | fuse.js     | fuse.common.js | fuse.esm.js              | fuse.esm.browser.js      |
```
Comment (sidvishnoi) [inline README.md]: I'm not understanding why `fuse.esm.js` might be needed. Bundlers can use `fuse.esm.browser.js`, right?

Comment (krisk) [reply README.md]: I added the explanation [here](https://github.com/krisk/Fuse/blob/bundler/dist/README.md)

Does it clarify things?

Comment (sidvishnoi) [reply README.md]: Not quite. I looked at both [`dist/fuse.esm.js`](https://github.com/krisk/Fuse/blob/5d76a678a21c7df1ddd11d2290afd3055a599fc9/dist/fuse.esm.js) and [`dist/fuse.esm.browser.js`](https://github.com/krisk/Fuse/blob/5d76a678a21c7df1ddd11d2290afd3055a599fc9/dist/fuse.esm.browser.js), there isn't much difference that could be used for tree shaking (there is a single default export in both). The only difference I saw is fuse.esm.js is transpiled, while browser one is not. When people are using bundlers, they often use transpilers along with it. So we shouldn't need a `fuse.esm.js` (and rename `fuse.esm.browsers.js` to `fuse.esm.js`), unless I'm missing something.

Comment (sidvishnoi) [reply README.md]: Slightly late to mention, but having a `fuse.esm.min.js` would be nice to have.
Thanks for working on it though! 😻

Comment (krisk) [reply README.md]: I can add it as part of v5.2.0

**Pull request merged**